proposals/0100-read-only-reserved-accounts.md

```diff
@@ -0,0 +1,140 @@
+write lock automatically demoted to a read-lock and their transactions still
+being executed. Instead, their transaction would fail and a lot of apps could be
+affected by this change in behavior.
+
+2. Demote write-locks lazily
+
+Another alternative is to demote write-locks while loading transaction accounts
+if a loaded account is owned by a reserved id like "native loader" or "sysvar".
+This approach means that a simpler set of reserved program ids could be tracked
+rather than tracking a full set of all reserved accounts.
+
+Again, the problem here is that the scheduler also needs to be aware of how
+write locks might be demoted in order to schedule transactions efficiently. So
+transaction schedulers would likely need to track a dynamic list of reserved
+accounts anyways to ensure that transaction locks are handled in exactly the
+same way as the runtime to avoid scheduling issues. Otherwise if a scheduler
+prematurely demotes transaction write locks before that sysvar exists, it could
+send an invalid batch of conflicting transactions to the runtime.
+
+## New Terminology
+
+Reserved Accounts: Any builtin program or sysvar or other account key managed by
+the Solana core protocol that should not be write locked by a transaction. This
+excludes the incinerator account which is designed to be write locked.
+
+## Detailed Design
+
+To determine the post-demotion set of write lock requested keys for a
+transaction, each transaction write locked account key is checked against a set
+of reserved keys maintained by the validator for a given block.
+
+The set of reserved keys will be initialized as the full set of sysvar and
+builtin program keys currently tracked in static lists in the Solana Labs
+validator implementation:
+
+sysvar_ids:
+
+- SysvarC1ock11111111111111111111111111111111
+- SysvarEpochSchedu1e111111111111111111111111
+- SysvarFees111111111111111111111111111111111
+- SysvarRecentB1ockHashes11111111111111111111
+- SysvarRent111111111111111111111111111111111
+- SysvarRewards111111111111111111111111111111
+- SysvarS1otHashes111111111111111111111111111
+- SysvarS1otHistory11111111111111111111111111
+- SysvarStakeHistory1111111111111111111111111
+- Sysvar1nstructions1111111111111111111111111
+
+builtin_program_ids:
+
+- Config1111111111111111111111111111111111111
+- Feature111111111111111111111111111111111111
+- NativeLoader1111111111111111111111111111111
+- Stake11111111111111111111111111111111111111
+- StakeConfig11111111111111111111111111111111
+- Vote111111111111111111111111111111111111111
+- 11111111111111111111111111111111
+- BPFLoader1111111111111111111111111111111111
+- BPFLoader2111111111111111111111111111111111
+- BPFLoaderUpgradeab1e11111111111111111111111
+
+The set of reserved keys may only be modified via feature gate activation. On
+epoch boundaries, validator implementations should add or remove reserved keys
+as dictated by feature gated code. The first feature gate activated modification
```

Comment:
How will/should accounts like those added in https://github.com/solana-labs/solana/pull/34624 be included? Presumably they ought to have write-locks demoted, but there will be a new one every epoch, which probably isn't feasible to feature-gate.

Reply:
I think it's ok if those accounts are able to be write locked for now.

Reply:
agreed, it's probably fine to take writelocks on these accounts. runtime will be doing its writing before transaction processing begins so there will be no contention there. 

---

the only reason i can think of to reference these accounts on chain would be to determine if an epoch rewards distribution is ongoing at execution time. this would require the ability to query block height, which we don't have. should that be made available in the future, we could add only the current epoch's epoch rewards partition metadata account address to the reserved set.

Reply:
We already have the epoch rewards sysvar available for checking if rewards distribution is happening so that case is covered.

Reply:
oh i thought we killed that. kinda... not necessary for anything other than `is_in_epoch_rewards_period()`, for which it's overkill

Reply:
@CriesofCarrots told me that the sysvar is sticking around

Reply:
Fwiw, I can imagine a scenario in which an on-chain program would want to get the partitioned rewards data from one of these accounts immediately on the epoch boundary in order to do something as soon as a reward is delivered, instead of waiting until the end of the distribution period. But it's not a super compelling use-case, and we can certainly enforce that such programs wait.

That said, my question was meant to be a general one. These new PDA sysvars aren't fundamentally different than statically addressed sysvars, and there may be others in the future that have a stronger on-chain use-case. Just kick the can?

Reply:
> [...] in order to do something as soon as a reward is delivered, instead of waiting until the end of the distribution period

Stake accounts can't be modified until the end of the distribution period so I don't think there's much value in doing anything immediately at the epoch boundary.

> These new PDA sysvars [...]

The PDAs you created for epoch rewards are stake program owned, aren't they? The stake program PDAs aren't much different from things like the Jito validator history program which creates a bunch of accounts to store historical validator performance. I don't think the epoch rewards PDA need to be enshrined as read-only at this point...

> Just kick the can?

Yup! We can address further write lock controls in a new SIMD for these types of use cases. This SIMD is for a few commonly used accounts that could disrupt a large chunk of transactions if they could be write locked. The approach relies on an in-memory set of accounts which is not designed to grow a lot.

Reply:
> Stake accounts can't be modified until the end of the distribution period so I don't think there's much value in doing anything immediately at the epoch boundary.

I'm a bit surprised you can look at the state of high-speed trading/speculation and make that argument. We're talking over a minute here.

> We can address further write lock controls in a new SIMD for these types of use cases.

Okay, let's do that :)

Reply:
> look at the state of high-speed trading/speculation

omg there are going to be options on newly distributed validator rewards :handshake: